File: 241. Different Ways to Add Parentheses/diffWaysToCompute.js
/**
 * @param {string} expression
 * @return {number[]}
 */
const diffWaysToCompute = (expression) => {
  const res = []
  const l = expression.length;

  if (l === 0) {
    return res;
  } 

  if (l === 1) {
    return [parseInt(expression)];
  }

  if (l === 2 && !isNaN(expression[0])) {
    return [parseInt(expression)];
  }

  for (let i = 0; i < l; i++) {
    const c = expression[i];
    if (!isNaN(c)) {
      continue;
    }

    const leftResult = diffWaysToCompute(expression.slice(0, i));
    const rightResult = diffWaysToCompute(expression.slice(i + 1));

    for (let left of leftResult) {
      for (let right of rightResult) {
        switch (c) {
          case '+':
            res.push(left + right);
            break;
          case '-':
            res.push(left - right);
            break;
          case '*':
            res.push(left * right);
            break;
        }
      }
    }
  }

  return res;
};


const expression = "2-1-1";
console.log('diffWaysToCompute: ', diffWaysToCompute(expression)); // [0, 2]

File: 241. Different Ways to Add Parentheses/diffWaysToCompute.py
from typing import List

class Solution:
  def diffWaysToCompute(self, expression: str) -> List[int]:
    res = []
    l = len(expression)
    if l == 0:
      return res
    
    if l == 1:
      return [int(expression)]
    
    if l == 2 and expression[0].isdigit():
      return [int(expression)]
    
    for i in range(l):
      if expression[i].isdigit():
        continue

      leftResult = self.diffWaysToCompute(expression[:i])
      rightResult = self.diffWaysToCompute(expression[i + 1:])

      for leftVal in leftResult:
        for rightVal in rightResult:
          if expression[i] == '+':
            res.append(leftVal + rightVal)
          elif expression[i] == '-':
            res.append(leftVal - rightVal)
          elif expression[i] == '*':
            res.append(leftVal * rightVal)
    
    return res


expression = "2-1-1"
solution = Solution()
print('diffWaysToCompute: ', solution.diffWaysToCompute(expression)) # [0, 2]
